### PR TITLE
Improve the error when SUPPORT_LONGJMP=0 but longjmp is used

### PR DESCRIPTION
**Pull request by kripken**
Before this change, the error would just be "undefined symbol: longjmp" which is
not that clear, and worse, if errors on undefined symbols were disabled, it would
just be a warning.

Also don't run the backend pass if longjmp support is disabled (not needed, but
seems more correct, and saves time).

### CODE REVIEW

File: src/library.js [hunk 1]
```diff
@@ -1766,19 +1766,27 @@ LibraryManager.library = {
...
-#if SUPPORT_LONGJMP
   longjmp__sig: 'vii',
+#if SUPPORT_LONGJMP
   longjmp: function(env, value) {
     _setThrew(env, value || 1);
     throw 'longjmp';
   },
+#else
+  longjmp__deps: [function() {
+    error('longjmp support was disabled (SUPPORT_LONGJMP=0), but it is required by the code (either set SUPPORT_LONGJMP=0, or remove uses of it in the project)');
```
Comment (sbc100) [inline src/library.js:1777]: The advice there should be to remove `SUPPORT_LONGJMP=0` right?  Since the default is `1`

**Comment by kripken** [comment]
(also added a change to not run the backend pass, see edited initial comment)

File: src/library.js [hunk 2]
```diff
@@ -1766,19 +1766,27 @@ LibraryManager.library = {
...
+  }],
+  // will never be emitted, as the dep errors at compile time
```
Comment (sbc100) [inline src/library.js:1779]: Interesting approach, although a bit magical

**Review by sbc100 — APPROVED** [review]

File: tests/test_other.py [hunk 3]
```diff
@@ -9452,3 +9452,14 @@ def test_SYSCALL_DEBUG(self):
...
+  def test_SUPPORT_LONGJMP_executable(self):
+    stderr = self.run_process([EMCC, path_from_root('tests', 'core', 'test_longjmp.c'), '-s', 'SUPPORT_LONGJMP=0'], stderr=PIPE, check=False).stderr
+    self.assertContained('error: longjmp support was disabled (SUPPORT_LONGJMP=0), but it is required by the code (either set SUPPORT_LONGJMP=0, or remove uses of it in the project)',
+                         stderr)
+
+  def test_SUPPORT_LONGJMP_object(self):
+    self.run_process([EMCC, path_from_root('tests', 'core', 'test_longjmp.c'), '-s', 'SUPPORT_LONGJMP=0', '-c', '-o', 'a.o'])
+    stderr = self.run_process([EMCC, 'a.o', '-s', 'SUPPORT_LONGJMP=0'], stderr=PIPE, check=False).stderr
+    self.assertContained('error: longjmp support was disabled (SUPPORT_LONGJMP=0), but it is required by the code (either set SUPPORT_LONGJMP=0, or remove uses of it in the project)',
+                         stderr)
```
Comment (sbc100) [inline tests/test_other.py:9465]: Do we need two different tests here?  I mean internally the object will be created either way.

Also SUPPORT_LONGJMP is ignored during compile phase so only needed at link time isn't it?

Comment (kripken) [reply tests/test_other.py:9466]: The PR makes the flag be noticed during compile time, as it doesn't send the pass to the backend when longjmp is disabled (i added this after the first comment, so maybe you viewed before that).

But yeah, the test does seem redundant as it is. I'll change it to cover a corner case of compiling with support but linking without.

File: src/library.js (hunk 1 above)
Comment (kripken) [reply src/library.js]: Right, thanks, I was off by 1...

File: src/library.js (hunk 2 above)
Comment (kripken) [reply src/library.js:1779]: (the other option would be to add the check somewhere in `jsifier.js` - this way it's right alongside the relevant `longjmp` logic, which seems better to me)

Comment (sbc100) [reply src/library.js:1779]: No I think this is fine.

File: tools/building.py [hunk 4]
```diff
@@ -438,8 +438,9 @@ def llvm_backend_args():
     allowed = ','.join(Settings.EXCEPTION_CATCHING_ALLOWED or ['__fake'])
     args += ['-emscripten-cxx-exceptions-allowed=' + allowed]
 
-  # asm.js-style setjmp/longjmp handling
-  args += ['-enable-emscripten-sjlj']
+  if Settings.SUPPORT_LONGJMP:
+    # asm.js-style setjmp/longjmp handling
+    args += ['-enable-emscripten-sjlj']
```
Comment (sbc100) [inline tools/building.py:443]: Doesn't this change mean that we would break projects that set `SUPPORT_LONGJMP` only at link time?   

Not sure that is great idea.  Also its nice to be able to make as many flags as possible into link-only flags.

Comment (sbc100) [reply tools/building.py:443]: Oh.. SUPPORT_LONGJMP=0 is the default so maybe not an issue?

Comment (kripken) [reply tools/building.py:443]: Yes, there is some risk of breakage, fair point. But OTOH if the flag is ignored at compile time, then just setting the flag at link time won't work for purposes of invokes - the object file will have been rewritten to use invokes already.

Maybe it's not worth making this change as proper invoke handling is landing any time now, but it does seem more correct to me, that if the flag is passed, we should assume there is no longjmp work to be done.

**Review by sbc100 — APPROVED** [review]
